src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -0,0 +1,129 @@
+
+  /**
+   * Checks for invalid cases (no user or tripKey).
+   * Gets the locations from datastore and prints them to writer.
+   */  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) 
+      throws IOException {
+    response.setContentType("application/json;");
+
+    // Initialize Datastore.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get current user.
+    Entity userEntity = AuthServlet.getCurrentUserEntity();
+
+    // Get tripKey.
+    String stringTripKey = request.getParameter(TRIP_KEY_PARAM);
+
+    // if no tripKey return to trips page.
+    if (stringTripKey == null) {
+      response.getWriter().println("No trip Key");
+      response.sendRedirect("/trips/");
+  
+    // if no user Entity return home.
+    } else if (userEntity == null) {
+      response.getWriter().println("No current User");
+      response.sendRedirect("/");
+    } else { 
+      Key tripKey = KeyFactory.stringToKey(stringTripKey);
+
+      // Gets the locations from datastore and writes them to .../get-map
+      doGetMap(response, datastore, userEntity, tripKey);
+    }
+  }
+
+  /**
+   * Gets the locations and prints them to writer.
+   */
+  public void doGetMap(HttpServletResponse response, 
+      DatastoreService datastore, Entity userEntity, Key tripEntityKey) throws IOException {
+  
+    // Get trip Entity based on trip key.
+    // Trip entity is needed to ensure that tripKey corresponds to a Trip under the current user.
+    Filter tripKeyFilter =
+      new FilterPredicate("__key__", FilterOperator.EQUAL, tripEntityKey);
+    Query tripQuery = new Query(Trip.TRIP, userEntity.getKey());
+    tripQuery.setFilter(tripKeyFilter);
+    PreparedQuery tripResults = datastore.prepare(tripQuery);
+    Entity tripEntity = tripResults.asSingleEntity();
+
+    // If no trip is found then redirect home
+    if (tripEntity == null) {
+      response.getWriter().println("No trip found");
+      response.sendRedirect("/");
+    } else {
+      // Get TripDay associated with the Trip.
+      // Post-MVP: change to select the desired tripDay.
+      Query tripDayQuery = new Query(TripDay.QUERY_STRING, tripEntity.getKey());
+      PreparedQuery tripDayResults = datastore.prepare(tripDayQuery);
+      Entity tripDayEntity = tripDayResults.asSingleEntity();
+
+      // Add origin as the first location.
+      List<String> locations = new ArrayList<>();
+      locations.add((String) tripDayEntity.getProperty(TripDay.ORIGIN));
+
+      // Add rest of POIs to locations list and write to writer.
+      Query locationsQuery = new Query(TripDay.LOCATION_ENTITY_TYPE, tripDayEntity.getKey());
+      locationsQuery.addSort(TripDay.ORDER);
+      PreparedQuery locationResults = datastore.prepare(locationsQuery); 
+      for (Entity locationEntity : locationResults.asIterable()) {
+        locations.add((String) locationEntity.getProperty(TripDay.NAME));
```

Comment:
It seems like you are just returning the tripday names? If that's the case, you should update the comment to reflect this.

But don't you need more information than this? Like the actual lat/lng location? I'm probably just not as familiar with this code, so could you explain it to me?

Reply:
The Maps API is able to route locations with String addresses. Since we're using autofill and suggested POIs to get user input, we are collecting valid String addresses as the user input, and I can just store/fetch these addresses from Datastore.